File: backend/src/main/java/med/doctor_connect/service/ScheduleService.java
package med.doctor_connect.service;

import med.doctor_connect.dto.CreateScheduleSlotRequest;
import med.doctor_connect.dto.ScheduleSlotDto;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

public interface ScheduleService {

    List<ScheduleSlotDto> getDoctorSchedules(UUID doctorId, LocalDate date);

    /**
     * Get only available (not fully booked) schedule slots for a doctor
     */
    List<ScheduleSlotDto> getAvailableSchedules(UUID doctorId, LocalDate date);

    ScheduleSlotDto createScheduleSlot(UUID doctorId, CreateScheduleSlotRequest request);

    ScheduleSlotDto updateScheduleSlot(UUID doctorId, UUID slotId, CreateScheduleSlotRequest request);

    void deleteScheduleSlot(UUID doctorId, UUID slotId);
}

File: frontend/doctor-connect-ui/src/app/services/schedule.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ScheduleSlot, CreateScheduleSlotRequest } from '../models';

@Injectable({
  providedIn: 'root'
})
export class ScheduleService {
  private readonly API_URL = 'http://localhost:8000/api/v1';

  constructor(private http: HttpClient) {}

  getDoctorSchedules(doctorId: string, date?: string): Observable<ScheduleSlot[]> {
    let params = new HttpParams();
    if (date) params = params.set('date', date);

    return this.http.get<ScheduleSlot[]>(`${this.API_URL}/doctors/${doctorId}/schedules`, { params });
  }

  createScheduleSlot(doctorId: string, request: CreateScheduleSlotRequest): Observable<ScheduleSlot> {
    return this.http.post<ScheduleSlot>(`${this.API_URL}/doctors/${doctorId}/schedules`, request);
  }

  updateScheduleSlot(doctorId: string, slotId: string, request: CreateScheduleSlotRequest): Observable<ScheduleSlot> {
    return this.http.put<ScheduleSlot>(`${this.API_URL}/doctors/${doctorId}/schedules/${slotId}`, request);
  }

  deleteScheduleSlot(doctorId: string, slotId: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/doctors/${doctorId}/schedules/${slotId}`);
  }

  // Public endpoint - Get available (not fully booked) schedule slots
  getAvailableSchedules(doctorId: string, date?: string): Observable<ScheduleSlot[]> {
    let params = new HttpParams();
    if (date) params = params.set('date', date);

    return this.http.get<ScheduleSlot[]>(`${this.API_URL}/doctors/${doctorId}/schedules/available`, { params });
  }
}
